commit aba2ffd06cdfba59cbc7e2d64973e9926f11b59c (refs/changes/90/1190/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-12-28T16:58:59-08:00 (2 years, 9 months ago)
    
    joyent/node-cueball#67 want way to turn off Error.captureStackTrace calls

diff --git a/lib/connection-fsm.js b/lib/connection-fsm.js
index cace7d3..a3aab67 100644
--- a/lib/connection-fsm.js
+++ b/lib/connection-fsm.js
@@ -151,7 +151,12 @@ ConnectionFSM.prototype.claim = function (stack, cb) {
 	mod_assert.func(cb, 'callback');
 	if (stack === undefined) {
 		var e = {};
-		Error.captureStackTrace(e);
+		if (mod_utils.stackTracesEnabled()) {
+			Error.captureStackTrace(e);
+		} else {
+			e.stack = 'Error\n at unknown (stack traces disabled)' +
+			    '\n at unknown (stack traces disabled)\n';
+		}
 		stack = e.stack;
 	}
 	this.cf_claimStack = stack.split('\n').slice(1).
@@ -188,7 +193,12 @@ ConnectionFSM.prototype.release = function (cb) {
 	    'connection is not held');
 
 	var e = {};
-	Error.captureStackTrace(e);
+	if (mod_utils.stackTracesEnabled()) {
+		Error.captureStackTrace(e);
+	} else {
+		e.stack = 'Error\n at unknown (stack traces disabled)\n' +
+		    ' at unknown (stack traces disabled)\n';
+	}
 	this.cf_releaseStack = e.stack.split('\n').slice(1).
 	    map(function (l) { return (l.replace(/^[ ]*at /, '')); });
 	this.once('stateChanged', function (st) {
diff --git a/lib/index.js b/lib/index.js
index 3e584cd..0ea7dde 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -12,6 +12,7 @@ const mod_resolver = require('./resolver');
 const mod_pmonitor = require('./pool-monitor');
 const mod_errors = require('./errors');
 const mod_cset = require('./set');
+const mod_utils = require('./utils');
 
 module.exports = {
 	HttpAgent: mod_agent.HttpAgent,
@@ -23,6 +24,9 @@ module.exports = {
 	resolverForIpOrDomain: mod_resolver.resolverForIpOrDomain,
 
 	poolMonitor: mod_pmonitor.monitor,
+	enableStackTraces: function () {
+		mod_utils.stackTracesEnabled.DEFAULT = true;
+	},
 
 	ClaimTimeoutError: mod_errors.ClaimTimeoutError,
 	NoBackendsError: mod_errors.NoBackendsError,
diff --git a/lib/pool.js b/lib/pool.js
index e4120bf..f312bb6 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -716,7 +716,12 @@ CueBallConnectionPool.prototype.claim = function (options, cb) {
 	}
 
 	var e = {};
-	Error.captureStackTrace(e);
+	if (mod_utils.stackTracesEnabled()) {
+		Error.captureStackTrace(e);
+	} else {
+		e.stack = 'Error\n at unknown (stack traces disabled)\n' +
+		    ' at unknown (stack traces disabled)\n';
+	}
 
 	/* If there are idle connections sitting around, take one. */
 	while (this.p_idleq.length > 0) {
diff --git a/lib/utils.js b/lib/utils.js
index 28ff071..ae8baae 100644
--- a/lib/utils.js
+++ b/lib/utils.js
@@ -10,11 +10,37 @@ module.exports = {
 	shuffle: shuffle,
 	planRebalance: planRebalance,
 	assertRecovery: assertRecovery,
-	assertRecoverySet: assertRecoverySet
+	assertRecoverySet: assertRecoverySet,
+	stackTracesEnabled: stackTracesEnabled
 };
 
 const mod_assert = require('assert-plus');
 
+stackTracesEnabled.DEFAULT = false;
+var mod_dtrace, dtProvider, dtProbe;
+
+function stackTracesEnabled() {
+	if (mod_dtrace === null)
+		return (stackTracesEnabled.DEFAULT);
+	if (mod_dtrace === undefined) {
+		try {
+			mod_dtrace = require('dtrace-provider');
+		} catch (e) {
+			mod_dtrace = null;
+			return (stackTracesEnabled.DEFAULT);
+		}
+		dtProvider = mod_dtrace.createDTraceProvider('cueball');
+		dtProbe = dtProvider.addProbe('capture-stack', 'int');
+		dtProvider.enable();
+	}
+	var en = stackTracesEnabled.DEFAULT;
+	dtProbe.fire(function () {
+		en = true;
+		return ([1]);
+	});
+	return (en);
+}
+
 function assertRecoverySet(obj) {
 	mod_assert.object(obj, 'recovery');
 	var keys = Object.keys(obj);
diff --git a/package.json b/package.json
index ca476eb..18c409e 100644
--- a/package.json
+++ b/package.json
@@ -17,6 +17,9 @@
     "vasync": ">=1.6.3 <2.0.0",
     "verror": ">=1.6.1 <2.0.0"
   },
+  "optionalDependencies": {
+    "dtrace-provider": "~0.8"
+  },
   "devDependencies": {
     "jsprim": ">=1.3.0 <2.0.0",
     "tape": ">=4.4.0 <5.0.0",
